### PR TITLE
Use Leafwing-Studios/cargo-cache in CI

### PR DESCRIPTION
**Pull request by BD103**
# Objective

- Our CI caches the `~/.cargo` and `target` folders, to take advantage of faster, incremental, compilation.
- The use of caches is quite inconsistent, with different formats for keys. It also does not take advantage of cache fallback (using an old cache if the current one does not exist).
- Caches are being created in the merge queue, but can never be used.
  - To verify this, go to any [validation job](https://github.com/bevyengine/bevy/actions/runs/8777168881) and check the `actions/cache@v4` step. It will always say `Cache not found for input keys: ...`.

## Solution

- Switch to using [Leafwing-Studios/cargo-cache](https://github.com/Leafwing-Studios/cargo-cache), which automatically handles cache keys, fallback, dependency updates, and toolchain updates.
- Caches are no longer used in the merge queue.
  - Github Actions has something called [cache isolation](https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows#restrictions-for-accessing-a-cache). Long story short, caches cannot be used if they were not created from the same branch or base branch.
  - For instance, a cache created on `main` can be used by any feature branch, but a cache created by `feature-a` branch cannot be used by the unrelated `feature-b` branch.

---

## Changelog

- All CI caching is now done with [Leafwing-Studios/cargo-cache](https://github.com/Leafwing-Studios/cargo-cache).
- The CI no longer creates inaccessible caches for validation jobs.

### CODE REVIEW

**Comment by BD103** [comment]
> You might be interested in this: https://github.com/Leafwing-Studios/cargo-cache

Thank you! I'll definitely take inspiration from that. I chose not to hash the lockfile because I figured that the amount of dependency change in 1 day cannot be so great that it would hurt performance, but that assumption will definitely require testing. It's good to see an alternate approach, though!

**Comment by BD103** [comment]
So far things are working well. I'm having some issues with the cache not being found, so I may have to move my testing over to my fork, where I can actually delete caches and cancel runs.

Building on MacOS and docs are the two longest jobs. I'm going to see if I can shorten their duration, if possible.

**Comment by BD103** [comment]
Following @TimJentzsch's interesting link, I've decided to switch this PR to use Leafwing-Studios/cargo-cache for everything. I think it's a lot more maintainable than my custom solution (and it's easy to use, too!).

Performance seems roughly the same, except for `run-examples-macos-metal` which didn't use caching before. It's hard to tell, since Github has two different MacOS runners with 3 and 4 CPU cores.

**Comment by TimJentzsch** [comment]
Previously attempted in #7627, but the action has also improved since then

**Comment by BD103** [comment]
> Previously attempted in #7627, but the action has also improved since then

Is there a reason the original PR was closed?

**Comment by TimJentzsch** [comment]
> > Previously attempted in #7627, but the action has also improved since then
> 
> Is there a reason the original PR was closed?

Mainly because it didn't get any more reviews and I was closing out stale PRs that I had still open.
I also found it hard to measure the impact of the change, it's hard to debug if a recompile is needed or unnecessary IMO

**Comment by TimJentzsch** [comment]
But I'm quite confident that the current caching solution is flawed, causing the cache to be outdated quite quickly, and that this PR will be an improvement.

I'm not entirely sure how well the action will work with as many different workflows and branches as we have in Bevy, the maximum cache side could be an issue.
But at least in smaller projects (e.g. Emergence) we have used this action successfully and reduced CI times significantly.

**Comment by BD103** [comment]
I feel like the risk is acceptable. Leafwing Studios is a trusted third-party, and the argument against using it is the same as any other Github Action we depend on. What if `dtolnay/rust-toolchain`, Super Linter, Typos, `cargo-release`, or any other action we use started downloading all tokens and force-pushing to main?

While we cannot guarantee that our dependencies will never run untrusted code, we can ensure that that untrusted code will not be able to do anything harmful by hardening our security and restricting permissions.

This is not an argument to remove those dependencies. I think they provide too much value to our build system to remove them entirely. If security is a worry we can fork the actions, use them instead in CI, and occasionally merge upstream changes.

**Comment by mockersf** [comment]
> the maximum cache side could be an issue.

This maximum?
Screenshot 2024-04-27 at 10 49 55

We almost never go below 50GB, sometimes over 100GB, GitHub seems... ok with that? That means we never have a cache that is more than one day old.

I think I would prefer an approach where we create cache only from the main branch, those should be available to every job

**Comment by TimJentzsch** [comment]
> I think I would prefer an approach where we create cache only from the main branch, those should be available to every job

What would you think about handling this in a separate issue/PR?
I think it could be tackled independently of the way we generate the cache and has its own up- and downsides that we could discuss seperately

**Review by janhohenheim — APPROVED** [review]

**Comment by BD103** [comment]
If / when this gets merged, it would be good to wipe all of the existing caches [here](https://github.com/bevyengine/bevy/actions/caches). This will make the migration a little quicker, and hopefully show the performance improvements sooner.

**Comment by alice-i-cecile** [comment]
Swapping to Controversial; I want to get Francois's sign-off here.

**Comment by janhohenheim** [comment]
Update to the newly released `v2` before merging.

**Comment by TimJentzsch** [comment]
> I think I would prefer an approach where we create cache only from the main branch, those should be available to every job

We can now do this via the `save-if` option, like this:

``yaml
- uses: Leafwing-Studios/cargo-cache@b46a3eb6f97357ee4bada528857bfff77d08b964
  with:
    save-if: ${{ github.ref == format('refs/heads/{0}', github.event.repository.default_branch) }}
```

**Comment by BD103** [comment]
Done!

@mockersf, could you please review this? You're the last step before this can be merged :)

**Comment by mockersf** [comment]
> @mockersf, could you please review this? You're the last step before this can be merged :)

I was waiting for my previous review to be addressed! Now unless there's an urgent need I'll do it after the Jam 🙂

File: .github/workflows/validation-jobs.yml
Comment (mnmaita) [inline .github/workflows/validation-jobs.yml:61]: Aren't these all parallel jobs and thus each of them will still need to retrieve cache? Or is there something I'm missing here? In other words, I'm questioning these removals as they aren't being replaced with the `cargo-cache` action but just being removed.

Comment (BD103) [reply .github/workflows/validation-jobs.yml:61]: Though it has been a while, I believe my reasoning was two-parted:

1. We have a cache storage limit of 10 GB, and Bevy builds take between 500 MB and 3 GB. We don't need validation checks to be fast, since they ideally only happen once, so its more valuable to make space for the fast-iteration checks run on every PR update.
2. Caches _can_ add a layer of instability. Since validation checks are the "final shield" between a PR and being merged, we probably don't want to risk it.

Comment (mnmaita) [reply .github/workflows/validation-jobs.yml:61]: Gotcha! So with that clarified, is there a reason why `build-and-install-on-iOS` still uses cache while the rest of the validation jobs won't anymore? Sorry if these were previously clarified in this thread, but I can't find any context on this. Does this have to do with what you said here?

> Building on MacOS and docs are the two longest jobs. I'm going to see if I can shorten their duration, if possible.

Thanks for answering and for your patience!

Comment (BD103) [reply .github/workflows/validation-jobs.yml:61]: > is there a reason why `build-and-install-on-iOS` still uses cache while the rest of the validation jobs won't anymore?

I think that was a mistake, let me remove it!

**Comment by BD103** [comment]
Closing due to inactivity. If someone wants to adopt this, feel free!

**Pull request closed without merging**
